lib/core/src/animation/physics.rs

```diff
@@ -0,0 +1,248 @@
+#[derive(Clone)]
+pub struct Properties {
+    data : Rc<RefCell<PropertiesData>>
+}
+
+impl Properties {
+    pub fn new
+    (kinematics:KinematicProperties, spring:SpringProperties, drag:DragProperties) -> Self {
+        let data = Rc::new(RefCell::new(PropertiesData::new(kinematics, spring, drag)));
+        Self { data }
+    }
+
+    pub fn kinematics(&self) -> KinematicProperties { self.data.borrow().kinematics.clone() }
+
+    pub fn set_kinematics(&mut self, kinematics:KinematicProperties) {
+        self.data.borrow_mut().kinematics = kinematics
+    }
+    pub fn spring(&self) -> SpringProperties { self.data.borrow().spring.clone() }
+
+    pub fn set_spring(&mut self, spring:SpringProperties) {
+        self.data.borrow_mut().spring = spring
+    }
+    pub fn drag(&self) -> DragProperties { self.data.borrow().drag.clone() }
+
+    pub fn set_drag(&mut self, drag:DragProperties) {
+        self.data.borrow_mut().drag = drag
+    }
+}
+
+// ========================
+// === PhysicsSimulator ===
+// ========================
+
+/// A 60 steps per second physics simulator used to simulate `Properties`.
+pub struct PhysicsSimulator {
+    _animator : Animator
+}
+
+/// Simulate the `KinematicProperties`.
+fn simulate_kinematics(properties:&mut KinematicProperties, dt:f32) {
+    properties.set_velocity(properties.velocity() + properties.acceleration() * dt);
+    properties.set_position(properties.position() + properties.velocity()     * dt);
+}
+
+/// Simulate dragging on `KinematicProperties`.
+fn simulate_dragging(kinematics:&mut KinematicProperties, drag:&DragProperties) {
+    let velocity = kinematics.velocity();
+    let speed    = velocity.norm();
+    kinematics.set_velocity(velocity / (1.0 + drag.amount() * speed));
+}
+
+/// Simulate spring on `KinematicProperties` attached to a fixed point.
+fn simulate_spring(properties:&mut KinematicProperties, spring_properties:&SpringProperties) {
+    let delta     = spring_properties.fixed_point() - properties.position();
+    let delta_len = delta.magnitude();
+    if delta_len > 0.0 {
+        let force_val = delta_len * spring_properties.coefficient();
+        let force     = delta.normalize() * force_val;
+        properties.set_acceleration(force / spring_properties.mass());
+    }
+}
+
+impl PhysicsSimulator {
+    /// Simulates `Properties` on `object`.
+    pub fn new<T>(mut object:T, properties:Properties) -> Self
+    where T: HasPosition + 'static {
+        let steps_per_second = 60.0;
+        let _animator = Animator::new(steps_per_second, move |delta_time| {
+            properties.kinematics().set_position(object.position());
+            simulate_spring(&mut properties.kinematics(), &properties.spring());
+            simulate_dragging(&mut properties.kinematics(), &properties.drag());
+            simulate_kinematics(&mut properties.kinematics(), delta_time);
+            object.set_position(properties.kinematics().position());
```

Comment:
Hmm, everything here seems really nicely decoupled (that is super cool)m, but I reall dont understand why you keep so many objects behind `RefCell`. If you could design that as keeping the ownership, this is **always** better way to do. I believe that spring properties and drag properties could be part of physics engine - then it will own that and ou will not use RefCells at all. Basically lets agree that using RefCells is bad and we should do it only if we need and ponly if we want to get nicer API. I believe that wihtout RefCells this user-facing API will be the same, am I right?

Reply:
Yeah, the bunch of RefCells were refactored out of the code as we discussed on Discord. The API is nicer too.